huaweicloud/services/apig/resource_huaweicloud_apig_appcode.go

```diff
@@ -0,0 +1,169 @@
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				ForceNew:    true,
+				Description: "The APPCODE value (content).",
+			},
+			"created_at": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The creation time of the APPCODE.",
+			},
+		},
+	}
+}
+
+func resourceAppcodeCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var (
+		resp *applications.AppCode
+
+		cfg    = meta.(*config.Config)
+		region = cfg.GetRegion(d)
+
+		instanceId = d.Get("instance_id").(string)
+		appId      = d.Get("application_id").(string)
+		appCode    = d.Get("value").(string)
+	)
+	client, err := cfg.ApigV2Client(region)
+	if err != nil {
+		return diag.Errorf("error creating APIG v2 client: %s", err)
+	}
+
+	if appCode != "" {
+		opt := applications.AppCodeOpts{
+			AppCode: appCode,
+		}
+		resp, err = applications.CreateAppCode(client, instanceId, appId, opt).Extract()
+		if err != nil {
+			return diag.Errorf("generating APPCODE failed: %s", err)
+		}
+	} else {
+		resp, err = applications.AutoGenerateAppCode(client, instanceId, appId).Extract()
+		if err != nil {
+			return diag.Errorf("auto generating APPCODE failed: %s", err)
+		}
+	}
+
+	d.SetId(resp.Id)
+
+	return resourceAppcodeRead(ctx, d, meta)
+}
+
+func resourceAppcodeRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var (
+		cfg    = meta.(*config.Config)
+		region = cfg.GetRegion(d)
+
+		instanceId = d.Get("instance_id").(string)
+		appId      = d.Get("application_id").(string)
+		appCodeId  = d.Id()
+	)
+	client, err := cfg.ApigV2Client(region)
+	if err != nil {
+		return diag.Errorf("error creating APIG v2 client: %s", err)
+	}
+
+	resp, err := applications.GetAppCode(client, instanceId, appId, appCodeId).Extract()
+	if err != nil {
+		return common.CheckDeletedDiag(d, err, fmt.Sprintf("error querying APPCODE (%s) from specified application (%s)",
+			appCodeId, appId))
+	}
+	mErr := multierror.Append(nil,
+		d.Set("region", region),
+		d.Set("value", resp.Code),
+		d.Set("created_at", resp.CreateTime),
+	)
+	if err = mErr.ErrorOrNil(); err != nil {
+		return diag.Errorf("error saving APPCODE resource fields: %s", err)
+	}
+	return nil
+}
+
+func resourceAppcodeDelete(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var (
+		cfg    = meta.(*config.Config)
+		region = cfg.GetRegion(d)
+
+		instanceId = d.Get("instance_id").(string)
+		appId      = d.Get("application_id").(string)
+		appCodeId  = d.Id()
+	)
+	client, err := cfg.ApigV2Client(region)
+	if err != nil {
+		return diag.Errorf("error creating APIG v2 client: %s", err)
+	}
+
+	err = applications.RemoveAppCode(client, instanceId, appId, appCodeId).ExtractErr()
+	if _, ok := err.(golangsdk.ErrDefault404); ok || err == nil {
+		return nil
+	}
+	return diag.Errorf("error deleting APPCODE (%s) from the specified application (%s): %s", appCodeId, appId, err)
```

Comment:
```suggestion
	if err != nil {
		if _, ok := err.(golangsdk.ErrDefault404); ok {
			return nil
		}
		return diag.Errorf("error deleting APPCODE (%s) from the specified application (%s): %s", appCodeId, appId, err)
	}
	return nil
```

Reply:
the original logic is clearer.
And in there, maybe `CheckDeletedDiag()` is better

Reply:
refactored